src/System.Private.CoreLib/src/System/GC.cs

```diff
@@ -657,9 +665,59 @@ internal static ulong GetSegmentSize()
             return RuntimeImports.RhGetGCSegmentSize();
         }
 
-        internal static T[] AllocateUninitializedArray<T>(int length)
+        internal static unsafe T[] AllocateUninitializedArray<T>(int length)
+        {
+            if (RuntimeHelpers.IsReferenceOrContainsReferences<T>()
+#if !DEBUG
+            // Small arrays are allocated using `new[]` as that is generally faster.
+            //
+            // The threshold was derived from various simulations. 
+            // As it turned out the threshold depends on overal pattern of all allocations and is typically in 200-300 byte range.
+            // The gradient around the number is shallow (there is no perf cliff) and the exact value of the threshold does not matter a lot.
+            // So it is 256 bytes including array header.
+            || (Unsafe.SizeOf<T>() * length < 256 - 3 * IntPtr.Size)
+#endif
+            )
+            {
+                return new T[length];
+            }
+
+            var pEEType = EETypePtr.EETypePtrOf<T[]>();
+
+            if (length < 0)
```

Comment:
The length check happens earlier on on the CoreCLR side. Can we end up with a different exception if we take the `return new T[length];` path above?